### ISSUE TITLE
Auth middleware example from docs does not work

### ISSUE THREAD

**Issue by StefanVDWeide** [nuxt-modules/supabase#129]
### Version
@nuxtjs/supabase: 0.3.0
nuxt: 3.0.0

### Steps to reproduce
Create an `auth.ts` file in the `middleware` folder as described here: https://supabase.nuxtjs.org/usage/composables/use-supabase-user#auth-middleware

### What is Expected?
For a user to be redirected to the correct page after logging in.

### What is actually happening?
When a user logs in, no redirect happens because the user object doesn't exist when using the`useSupabaseUser()` in the `auth.ts` file. This means the check is is always false and the user is redirect to the `/login` page.

**Comment by timdiacon** [comment]
We're running into problems with this too having updated to supabase v2.

Previously when using supabase v1 our middleware worked as per the example in docs. If we log the value from `useSupabaseUser()` we can see it is as expected in the terminal logs / SSR but on the client it's logged as undefined?

**Comment by JeffValadares** [comment]
I noticed that when we call the supabase login function it creates a session in the local storage and only after a lot of time it creates the cookie session and the the user object gets populated.

**Comment by deixadilson** [comment]
I noticed that when we call the supabase login function it creates a session in the local storage and only after a lot of time it creates the cookie session and the the user object gets populated.

**Comment by larbish** [comment]
I can not reproduce in local. On which platform are you experiencing this ? Can you please create a reproduction ?

**Comment by deixadilson** [comment]
Mine is hosted in netlify
https://stalwart-marshmallow-429f9d.netlify.app/

This is the repository: https://github.com/JeffValadares/CoEnergie

**Comment by deixadilson** [comment]
Mine is hosted on netlify
https://stalwart-marshmallow-429f9d.netlify.app/

This is the repository: https://github.com/JeffValadares/CoEnergie

**Comment by larbish** [comment]
Closing in favor of #101

**Comment by deixadilson** [comment]
This also happens on my localhost.

The session is set only in local storage and the user is not redirected.
Only after a long wait the session is stored in the cookie and then the user is redirected.

On netlify the user s never redirected.

**Comment by Aybee5** [comment]
I am facing a similar issue even on local development, when I signing with a password and am redirected to the auth page, the user cookies are not set, till I switched to another tab and come back

Here is how I am login in 

```js
async function loginUser() {
  try {
    await auth.signInWithPassword({
      email: email.value,
      password: password.value,
    });
    return navigateTo("/dashboard");
  } catch (error) {
    console.log({ error });
  }
}
```

**Comment by Wafje** [comment]
I fixed it by using `useSupabaseAuthClient()` instead of `useSupabaseClient()` to sign in and out.

**Comment by Aybee5** [comment]
Wow it works, but that's strange, I also looked inside the package and saw
the implementation should work correctly but since it's not working I
thought something must be causing that.

If we are not supposed to use the  `useSupabaseClient`, then I believe it
should be communicated or the auth inside should be removed for the v2

**Comment by deixadilson** [comment]
> I fixed it by using `useSupabaseAuthClient()` instead of `useSupabaseClient()` to sign in and out.

It works! Thank you!

**Comment by StefanVDWeide** [comment]
Strangly enough this does not solve the issue for me locally. I tried creating a minimal example but you can't really do this without running Supabase itself, otherwise there is nothing to login to. 

My setup is really simple and I fail to see what it incorrect. My login component looks like:

``` ts 
const authClient = useSupabaseAuthClient();
const router = useRouter();

// Local state
const userEmailLogin = ref("");
const userPasswordLogin = ref("");

// Methods
// Login method using email
const login = async () => {
    const { error } = await authClient.auth.signInWithPassword({
        email: userEmailLogin.value,
        password: userPasswordLogin.value
    })
    if (error) {
        return alert('Something went wrong during login!')
    }
    router.push("/dashboard");
}
```
And the middleware file is exactly the same as in the official example in the docs:

``` ts
export default defineNuxtRouteMiddleware((_to, _from) => {
    const user = useSupabaseUser();

    if (!user.value) {
        return navigateTo("/login");
    }
})

```

I don't think we should close this issue just yet. Any ideas what this could be?

Thanks!

**Comment by tungquach** [comment]
Use `useSupabaseAuthClient` fix the problem but one more problem that when I refresh the page `useSupabaseUser()` return `null`

```
const user = useSupabaseUser()
console.log(user.value) // returns null
```

**Comment by Wafje** [comment]
@StefanVDWeide

can you try this instead of `router.push`?
```
const user = useSupabaseUser();
watchEffect(async () => {
  if (user.value) {
    navigateTo("/dashboard");
  }
});
```

That is how it is working on my end.

**Comment by ahku** [comment]
Same here, I'm using the useSupabaseAuthClient to login and logout, and it works locally. But on Netlify it doesn't.

**Comment by deixadilson** [comment]
Locally the cookie is set and the user is being redirected, only with useSupabaseAuthClient
On Netlify I need to seting the cookie manually with the code below
```
      const { data: response, error } = await supabase.auth.signInWithPassword(formData);
      if (error) throw error;
      const accessToken = useCookie('sb-access-token');
      const refreshToken = useCookie('sb-refresh-token');
      accessToken.value = response.session?.access_token ?? null;
      refreshToken.value = response.session?.refresh_token ?? null;
```

**Comment by deixadilson** [comment]
You can try to set the cookie manually with the code below:

```
      const { data: response, error } = await supabase.auth.signInWithPassword(formData);
      const accessToken = useCookie('sb-access-token');
      const refreshToken = useCookie('sb-refresh-token');
      accessToken.value = response.session?.access_token ?? null;
      refreshToken.value = response.session?.refresh_token ?? null;
```

**Comment by brandondtaylor01** [comment]
https://github.com/brandondtaylor01/nuxt3-supabase-bad-sessions
Here's a small repo that I created demoing the issue that I am seeing. Several weird things happening:

1. sb-access and sb-refresh aren't created until the window is minimized / tabbed out and back in.
2. auth.ts middleware doesn't know about the user after successful signin (the SIGNED_IN event has fired, and only then navigation occurs)
3. Signing out via the signOut method does not remove the sb-access and sb-refresh cookies. This causes the middleware to "remember" the user, despite not being currently signed in, and the local storage is cleared. Deleting these cookies manually fixes that.

I would love to have a solution to this. At this point, I suppose I'll just use useSupabaseAuthClient

**Comment by larbish** [comment]
@brandondtaylor01 For authentification purpose (`client.auth` methods) , you should use the `useSupabaseAuthClient`. Then you can fetch your data with the `useSupabaseClient`.

**Comment by brandondtaylor01** [comment]
> @brandondtaylor01 For authentification purpose (`client.auth` methods) , you should use the `useSupabaseAuthClient`. Then you can fetch your data with the `useSupabaseClient`.

I did also notice that, even when using that composable, the middleware still did not receive the data of the user until those cookies were created. And that still required a tab-out and tab-in to accomplish. I managed to get it to work by altering the middleware to this:

`export default defineNuxtRouteMiddleware(async (to) => {
  const client = useSupabaseAuthClient();

  if(!(await client.auth.getUser()).data.user) {
    return navigateTo('/sign-in/');
  }

  // no need to redirect, continue.
  // ...
});`

It may not be optimal, but it works.

**Comment by KeoneSomers** [comment]
I am also experiencing this issue. It does not occur when testing locally, only when hosted on Vercel.

**Comment by timdiacon** [comment]
Just to add to we're also experiencing no problems locally but useSupabaseUser is null when deployed both on Netlift and Vercel

**Comment by larbish** [comment]
The [demo](https://github.com/nuxt-modules/supabase/tree/main/demo) has been updated with the last version of both Nuxt and Supabase Module. It's deployed on Netlify.  It's using middleware as it is explained in the doc. Please let me know if you are still experiencing this issue on your own project. A reproduction is needed to help 🙏

**Comment by tungquach** [comment]
> The [demo](https://github.com/nuxt-modules/supabase/tree/main/demo) has been updated with the last version of both Nuxt and Supabase Module. It's deployed on Netlify.  It's using middleware as it is explained in the doc. Please let me know if you are still experiencing this issue on your own project. A reproduction is needed to help 🙏 

New version fixed problem, thank you so much
